sky/task.py

```diff
@@ -430,7 +431,10 @@ def from_yaml_config(
             mount_path = storage[0]
             assert mount_path, 'Storage mount path cannot be empty.'
             try:
-                storage_obj = storage_lib.Storage.from_yaml_config(storage[1])
+                resources = config.get('resources', None)
+                region = resources.get('region', None) if resources else None
+                storage_obj = storage_lib.Storage.from_yaml_config(
+                    storage[1], region)
```

Comment:
While this is a solution works, it breaks the abstractions we've drawn between `Storage` and `Store`. `Storage` is intended to be agnostic to the underlying `Store` (and consequently, the region the `Store` is in). As a result, I don't think `Storage` should not have a `region` field.

As you note, the bug occurs only when a `store` is specified in the YAML spec. This happens because we try to add the store at YAML parse:
https://github.com/skypilot-org/skypilot/blob/e794971548584cdc9c8d6c1bde4b734c151ef3c4/sky/data/storage.py#L1046-L1047

Instead, do you think we can delay this `add_store` from the YAML parsing to happen after the optimization step (see [execution.py::_execute](https://github.com/skypilot-org/skypilot/blob/e794971548584cdc9c8d6c1bde4b734c151ef3c4/sky/execution.py#L95) for the sequencing) so that it gets handled in the [sync_storage_mounts](https://github.com/skypilot-org/skypilot/blob/master/sky/task.py#L946-L960) step? At that step SkyPilot knows the region that the task will be scheduled in, so we can accurately specify the region to create the bucket in.

Reply:
I removed the `add_store` from YAML parsing and noticed that the buckets were still being created as the init function inside of `Storage` was adding to `sky_stores` (prev at storage.py, 527): 
```python
        if handle is not None:
            self.handle = handle
            # Reconstruct the Storage object from the global_user_state
            logger.debug('Detected existing storage object, '
                         f'loading Storage: {self.name}')
            self._add_store_from_metadata(self.handle.sky_stores)
``` 
The init function is triggered by the following in `from_yaml_config` (storage.py, 1043-1046):  
```python
        storage_obj = cls(name=name,
                          source=source,
                          persistent=persistent,
                          mode=mode)
``` 

After removing `self.add_store_from_metadata` from `Storage` init function, the bucket creation process happens after the Optimization step and creates the bucket in the specified region.

Reply:
Great! Can you also verify if `TestStorageWithCredentials`, `test_file_mounts`, `test_gcp_storage_mounts_with_stop` and `test_managed_jobs_storage` also pass?